Topic/tcp channel tests

d-e-s-o opened the pull request:
This change concludes the TCP channel infrastructure. Most importantly we run the existing tests for the observable-observer infrastructure to also test TCP channel based observers. Also, run the test in the GitLab CI/CD pipeline.

ryzhyk commented on rust/template/tcp_channel/receiver.rs:
```diff
@@ -25,15 +27,55 @@ use serde::de::DeserializeOwned;
+}
+
+#[derive(Debug)]
+struct Data {
```
`Data` is not a great name for this data structure. `ConnectionState` maybe?
Also, can't you just add the third option to `Fd` for closed state and not need this data type at all?

d-e-s-o replied:
Sure, I'll change the name. Yes, a third state may be possible. I'll try it out. I iterated through so many variants that when I finally had something working I did not think too hard to reduce the state even more.

d-e-s-o commented:
GitLab pipeline succeeded: https://gitlab.com/ddlog/differential-datalog-d-e-s-o/pipelines/87446935

ryzhyk commented on rust/template/tcp_channel/receiver.rs:
```diff
@@ -46,6 +88,10 @@ where
     /// assigned port (in the form of the full `SocketAddr`), use the
     /// `addr` method.
+    ///
+    /// Note that the receiver may require some time after creation
```
Why is that? You will have called `bind` by the time this method returns, so it should be ok for `TcpSenders` to connect. The connections will be processed once the listener thread calls `accept`. I might be wrong.

d-e-s-o replied:
Based on my observations my conclusion was that until `accept` is issued you'd still get a connection refused when trying to connect. The mere act of `bind`ing only prepares the socket and "reserves" a port. Or so that was my conclusion. I'll verify tomorrow.

ryzhyk replied:
From `TcpListener` documentation:

```
After creating a TcpListener by binding it to a socket address, it listens for incoming TCP connections. These can be accepted by calling accept or by iterating over the Incoming iterator returned by incoming.
```

What I think this means is that `TcpListener.bind()` internally calls `listen()` from `sys/socket.h`, which in turn means that the socket should already be queuing incoming connections.

BTW, I think there is nothing wrong with `TcpSender` retrying regardless of this; just trying to figure out what's going on...

d-e-s-o replied:
Yeah, I tried it out and it works as you suspected it would. I agree that eventually we may need a `retry_connect` but I guess for now we don't.

ryzhyk commented on rust/template/tcp_channel/receiver.rs:
```diff
@@ -122,14 +198,22 @@ where
+        self.thread
+            .take()
+            .map(JoinHandle::join)
+            .unwrap()
```
Won't this crash if the receiver thread returns an error?

d-e-s-o replied:
Yeah. For now I concluded that this is better than not having any signalling that something went wrong (which was the cause of a long debug session already). I'll change this to behave more friendly once I added logging.

ryzhyk replied:
As long as there is a big FIXME in the code :)

ryzhyk commented on test/datalog_tests/server_api/tests/deltas.rs:
```diff
@@ -1,58 +1,113 @@
+        // environment in a mutable way.
+        // We need an option because Vec::new is not const and so we can't
+        // use it to initialize a static.
+        static mut DELTAS: Option<Mutex<Vec<(usize, Record)>>> = None;
```
would `lazy_static` help here?

d-e-s-o replied:
I tried it, but the compiler threw some weird errors and my conclusion was that it does not deal with `mut` state, it's merely for initializing constants. But let me retry, it's possible I screwed up.

ryzhyk replied:
`DELTAS` does not have to be `mut`, does it? It's only `mut` here so you can initialize it. This is exactly what `lazy_static` is supposed to solve.

d-e-s-o replied:
It does have to mutable, because we accumulate the deltas in this vector. It works as long as there is no `mut` involved, but once I change the declaration the compiler bails out:
```
error: no rules expected the token `mut`
 --> src/main.rs:6:12
  |
6 |     static mut ref DELTAS: Mutex<Vec<usize>> = {
  |            ^^^ no rules expected this token in macro call
```

```
error: no rules expected the token `DELTAS`
 --> src/main.rs:6:20
  |
6 |     static ref mut DELTAS: Mutex<Vec<usize>> = {
  |                    ^^^^^^ no rules expected this token in macro call
```

d-e-s-o replied:
We clarified that in person. I'll change the code to use `lazy_static` :)

d-e-s-o commented:
I've addressed all review comment. New gitlab pipeline: https://gitlab.com/ddlog/differential-datalog-d-e-s-o/pipelines/87724508

Thanks for the review!
